src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = state;
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = combineProperties(previousState, action.extraData);
+        break;
+      case 'removeProperties':
+        previousState = removeProperties(previousState, action.keysToRemove);
+        break;
+      case 'clear':
+        previousState = clearState(previousState);
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    history.push(previousState);
+  }
+
+  return history;
+}
+
+function combineProperties(state, properties) {
+  return Object.assign({}, state, properties);
```

Comment:
The function name `combineProperties` does not clearly convey that it creates a new object. Consider renaming it to reflect that it creates a new state with added properties, such as `createStateWithAddedProperties`.